spec/memoize_query_spec.rb

```diff
@@ -443,11 +443,63 @@ def expect_to_eq_with_or_without_redis
     end
   end
 
-  it 'does not memoize queries with NOT' do
-    expect_not_to_use_redis do
-      Site.where(id: Site.where.not(a: 1)).to_a
+  context 'when the queries have NOT' do 
+    let!(:record) { Site.create!(a: 1, b: 1) }
+    let!(:relation1_with_only_not) { Site.where.not(a: 2) }
+    let!(:relation2_with_only_not) { Site.where.not(a: 2, b: 1) }
+    let!(:relation3_with_only_not) { Site.where.not(a: 2).where.not(b: 1) }
+    let!(:relation_with_not_and_other) { Site.where.not(a: 2).where(b: 1) }
+
+    it 'does not memoize queries with only NOT' do
+      expect_not_to_use_redis do
+        expect(relation1_with_only_not.to_a).to eq([record])
+      end
+
+      expect_not_to_use_redis do
+        expect(relation2_with_only_not.to_a).to eq([record])
+      end
+
+      expect_not_to_use_redis do
+        expect(relation3_with_only_not.to_a).to eq([])
+      end
     end
-  end
+
+    it 'memoizes queries with both NOT and other bound conditions' do
+      expect_to_use_redis do
+        expect(relation_with_not_and_other.to_a).to eq([record])
+      end
+    end
+
+    it 'it updates the affected query result' do
+      record.update(a: 2)
+
+      expect_to_use_redis do
+        expect(relation_with_not_and_other.reload.to_a).to eq([])
+      end
+    end
+
+    it 'only invalidates the affected query result sets' do
+      RedisMemo::Cache.with_local_cache do 
+        # WHERE a != 2 and b = 1
+        expect_to_use_redis do
+          expect(relation_with_not_and_other.to_a).to eq([record])
+        end
+        
+        Site.create!(b: 2)
+
+        # The new created record does not affect WHERE a != 2 and b = 1
+        expect_not_to_use_redis do
+          expect(relation_with_not_and_other.reload.to_a).to eq([record])
+        end
```

Comment:
If you reload the relation, it will force it to fetch from the db again right? How does this test that the other relation isn't affected by the new record?

Reply:
In this case, the versions will be the same because creating `Site.create!(b: 2)` won't affect the dependent versions. let me know if this answers your question.

Reply:
Ah got it, I see this is wrapped around `RedisMemo::Cache.with_local_cache`. Do you need the `reload` then?

Reply:
I don't necessarily need but want to make sure the result is up to date in this way in order to reflect the test purpose (since for the next test, we do need a reload to get the different result of the relation)